### PR TITLE
hotfix/t/WEB-2012_new_detail_transactions_friendly_name_bug: stop usi…

### PR DESCRIPTION
**Pull request by albertovillalobos**
…ng title to match html nodes to configs

Here's the bug fix.

@codyjca @Vicentecarrillo 
cc @mikeweaver @nburwell

### CODE REVIEW

File: dist/backdraft.js [hunk 1]
```diff
@@ -1275,8 +1275,19 @@ _.extend(Plugin.factory, {
 
     render : function() {
       var cells = this.findCells(), node;
-      _.each(this.columnsConfig, function(config) {
-        node = cells.filter(config.nodeMatcher(config));
+      var configsArray = this.columnsConfig;
+      var cleanConfigs = this.columnsConfig.map(function(a,b,c){
```
Comment (Vicentecarrillo) [inline dist/backdraft.js]: what the the parameters a,b,c? can you rename them to something more descriptive?

File: spec/plugins/data_table/server_side.js [hunk 2]
```diff
@@ -747,6 +747,67 @@ describe("DataTable Plugin", function() {
...
+
+      jasmine.Ajax.install();
+      mockResponse = new MockResponse();
+
+      table = new app.Views.T({ collection : collection });
+      table.render();
+      jasmine.Ajax.requests.mostRecent().response(mockResponse.get());
+      history.pushState(null, null, "_SpecRunner.html");
+      cg = table.configGenerator();
+    });
```
Comment (Vicentecarrillo) [inline spec/plugins/data_table/server_side.js]: not seeing any assertions in here?

File: dist/backdraft.js (hunk 1 above)
Comment (albertovillalobos) [inline dist/backdraft.js]: Addressed

File: spec/plugins/data_table/server_side.js (hunk 2 above)
Comment (albertovillalobos) [inline spec/plugins/data_table/server_side.js]: I wasn't sure what to assert here. Basically without the fix the test fails because an error is thrown.

Comment (nburwell) [inline spec/plugins/data_table/server_side.js]: I'd say to assert you have 2 filter views, one for cost and one for cost2 .. either soething that gets added to the dom, or that exists off of `cg`, ..

File: spec/plugins/data_table/server_side.js [hunk 3]
```diff
@@ -747,6 +747,68 @@ describe("DataTable Plugin", function() {
...
+
+      collection = new app.Collections.Col();
+
+      jasmine.Ajax.install();
+      mockResponse = new MockResponse();
+
+      table = new app.Views.T({ collection : collection });
+      table.render();
+      jasmine.Ajax.requests.mostRecent().response(mockResponse.get());
+      expect(table).toBeDefined();
```
Comment (Vicentecarrillo) [inline spec/plugins/data_table/server_side.js]: does this assertion guarantee that there aren't duplicate filter views?

Comment (albertovillalobos) [inline spec/plugins/data_table/server_side.js]: Yes, the test will fail when there are duplicate filters table will not be defined. On top of that the test will error out because of a duplicate filter view.

File: dist/backdraft.js [hunk 4]
```diff
@@ -1275,8 +1275,19 @@ _.extend(Plugin.factory, {
...
+        if (b===0) {
+          return a
+        }
+        if(a.attr&&a.visible){
```
Comment (nburwell) [inline dist/backdraft.js]: please match the styleguide for spacing around `if` and `{` etc, throughout this PR:

```
if (a.attr && a.visible) {
  ^       ^  ^          ^
```

File: dist/backdraft.js [hunk 5]
```diff
@@ -1275,8 +1275,19 @@ _.extend(Plugin.factory, {
 
     render : function() {
       var cells = this.findCells(), node;
-      _.each(this.columnsConfig, function(config) {
-        node = cells.filter(config.nodeMatcher(config));
+      var configsArray = this.columnsConfig;
+      var cleanConfigs = this.columnsConfig.map(function(a,b,c){
+        if (b===0) {
+          return a
```
Comment (nburwell) [inline dist/backdraft.js]: Please throw a `;` at end of all the returns

File: dist/backdraft.js [hunk 6]
```diff
@@ -1275,8 +1275,19 @@ _.extend(Plugin.factory, {
...
+        }
+        if(a.attr&&a.visible){
+          return a}
+        }).filter(function(n){
+           return n != undefined
```
Comment (nburwell) [inline dist/backdraft.js]: surprised jslint didn't squawk (do we not have it on backdraft?) but this should be `!==`.. or do we want `typeof n !== "undefined"` ?

File: dist/backdraft.js [hunk 7]
```diff
@@ -1275,8 +1275,19 @@ _.extend(Plugin.factory, {
 
     render : function() {
       var cells = this.findCells(), node;
-      _.each(this.columnsConfig, function(config) {
-        node = cells.filter(config.nodeMatcher(config));
+      var configsArray = this.columnsConfig;
```
Comment (nburwell) [inline dist/backdraft.js]: don't see this used? please remove

File: dist/backdraft.js [hunk 8]
```diff
@@ -1275,8 +1275,18 @@ _.extend(Plugin.factory, {
 
     render : function() {
       var cells = this.findCells(), node;
-      _.each(this.columnsConfig, function(config) {
-        node = cells.filter(config.nodeMatcher(config));
+      var cleanConfigs = this.columnsConfig.map(function(currentValue,index){
```
Comment (nburwell) [inline dist/backdraft.js]: What does `cleanConfigs` mean?  what are you cleaning up or normalizing? Can you describe it in the variable name so the code is more clear?

File: dist/backdraft.js [hunk 9]
```diff
@@ -1275,8 +1275,18 @@ _.extend(Plugin.factory, {
...
+        if (index===0) {
```
Comment (nburwell) [inline dist/backdraft.js]: why does the first item have special treatment?

File: dist/backdraft.js [hunk 10]
```diff
@@ -1275,8 +1275,18 @@ _.extend(Plugin.factory, {
...
+          return currentValue
+        }
+        if(currentValue.attr&&currentValue.visible){
```
Comment (nburwell) [inline dist/backdraft.js]: is this a net new change or no.. just trying to understand the implementation based on the code that is here... looks like now columns without an `attr` are skipped over, is that true? 

I think we have some tables that have columns with an `attr` key?  If my above assumption is true (and it may not be), why wouldn't we enforce `attr` as a required key up in the column config definition

File: dist/backdraft.js (hunk 8 above)
Comment (albertovillalobos) [inline dist/backdraft.js]: cleanConfigs is the array of column configs without any non-visible or undefined configs. I can't come up with a better name, `definedConfigs`, `availableConfigs` ?

File: dist/backdraft.js (hunk 10 above)
Comment (albertovillalobos) [inline dist/backdraft.js]: All columns should have attr except for the bulk column. So on this case I check that either they're bulk or have attr and are visible.

File: dist/backdraft.js (hunk 9 above)
Comment (albertovillalobos) [inline dist/backdraft.js]: It was meant for the bulk column but now I should check it it's bulk not if it's index 0.

File: dist/backdraft.js (hunk 7 above)
Comment (albertovillalobos) [inline dist/backdraft.js]: addressed

File: spec/plugins/data_table/server_side.js (hunk 2 above)
Comment (albertovillalobos) [inline spec/plugins/data_table/server_side.js]: Addressed

File: dist/backdraft.js (hunk 4 above)
Comment (albertovillalobos) [inline dist/backdraft.js]: Addressed

File: dist/backdraft.js (hunk 5 above)
Comment (albertovillalobos) [inline dist/backdraft.js]: Addressed

File: dist/backdraft.js (hunk 6 above)
Comment (albertovillalobos) [inline dist/backdraft.js]: Addressed

**Comment by albertovillalobos** [comment]
I've addressed all the CR, round 2?
@Vicentecarrillo 
cc @nburwell @mikeweaver

File: src/plugins/data_table/row.js [hunk 11]
```diff
@@ -10,8 +10,16 @@ var Row = (function() {
...
       var cells = this.findCells(), node;
-      _.each(this.columnsConfig, function(config) {
-        node = cells.filter(config.nodeMatcher(config));
+      var cleanConfigs = this.columnsConfig.map(function(currentValue,index){
+        if (currentValue.bulk || ((currentValue.attr || currentValue.title) && currentValue.visible) ) {
+          return currentValue;
+        }
+      }).filter(function(currentValue) {
+         return currentValue !== undefined;
+        }
```
Comment (Vicentecarrillo) [inline src/plugins/data_table/row.js]: is this filter necessary? not following how an undefined config would get passed back from the map function

Comment (albertovillalobos) [inline src/plugins/data_table/row.js]: Yes, the map only replaces stuff, doesn't splice. So it'd look like `[object, object, undefined, object, object, undefined]`, the filter removes the undefined parts.

File: dist/backdraft.js (hunk 10 above)
Comment (nburwell) [inline dist/backdraft.js]: But what if the first column is not bulk, but doesn't have an attr? You're blindly including the first column always.  Also the bulk column can be anywhere, not just the first position.. 

> All columns should have attr or title

Okay, that sounds reasonable, yes.. where is this code checking title though?

File: src/plugins/data_table/row.js (hunk 11 above)
Comment (mikeweaver) [inline src/plugins/data_table/row.js]: Is that true? I thought map ignored undefined results?

File: dist/backdraft.js (hunk 8 above)
Comment (mikeweaver) [inline dist/backdraft.js]: How about visibleColumnConfigs? I think that matches up with what you are doing here, trying to get the indexes of the visible columns.

Also, seems like this could be extracted into a function to make the code easier to read.

**Comment by albertovillalobos** [comment]
Alright, things changed a lot now. 
@Vicentecarrillo @codyjca CR please
cc @nburwell @mikeweaver

File: src/plugins/data_table/data_table.js [hunk 12]
```diff
@@ -433,6 +433,16 @@ var LocalDataTable = (function() {
         var title = this.textContent || this.innerText;
         var col = cg.columnConfigByTitle.attributes[title];
 
+        // Try to grab it by attribute, if possible
+        var matches = this.className.match(/(?:^|\s)column-(.*?)(?:\s|$)/);
```
Comment (mikeweaver) [inline src/plugins/data_table/data_table.js]: You could use column-([^\s]*) to make the regex a little safer. (It will not consume any spaces that way)

File: src/plugins/data_table/column_types/attr.js [hunk 13]
```diff
@@ -4,7 +4,7 @@ app.view.dataTable.columnType(function(columnType) {
   });
 
   columnType.nodeMatcher(function(config) {
-    return "." + Backdraft.Utils.toCSSClass(config.title);
+    return "." + Backdraft.Utils.toCSSClass(config.attr || config.title);
```
Comment (mikeweaver) [inline src/plugins/data_table/column_types/attr.js]: Why are we supporting both ways? Why not just attr?

Comment (albertovillalobos) [inline src/plugins/data_table/column_types/attr.js]: Some backdraft tables don't use config.attr at all.

Comment (nburwell) [inline src/plugins/data_table/column_types/attr.js]: ah, Mike's comment is a good one here though, I led you slightly astray yesterday Alberto.  Yes, definitely true that some tables don't use config.attr at all, and that is a main reason for the redo of the implementation here.. but this `column_types/attr.js` should only be used on columns where there **is** an attr. (See lines 2~4).

So in this file, I think it is correct / more accurate to just use config.attr on this line and line 16 below.

File: src/plugins/data_table/column_types/base.js [hunk 14]
```diff
@@ -4,7 +4,7 @@ app.view.dataTable.columnType(function(columnType) {
   });
 
   columnType.nodeMatcher(function(config) {
-    return "." + Backdraft.Utils.toCSSClass(config.title);
+    return "." + Backdraft.Utils.toCSSClass(config.attr || config.title);
```
Comment (nburwell) [inline src/plugins/data_table/column_types/base.js]: and this file can just stay as config.title because notice it's only used when `!config.attr` (per line 3)

File: src/plugins/data_table/column_types/base.js [hunk 15]
```diff
@@ -17,7 +17,7 @@ app.view.dataTable.columnType(function(columnType) {
       bSortable: sortable,
       bSearchable: searchable,
       sTitle: config.title,
-      sClass : Backdraft.Utils.toCSSClass(config.title),
+      sClass : Backdraft.Utils.toCSSClass(config.attr || config.title),
```
Comment (nburwell) [inline src/plugins/data_table/column_types/base.js]: ditto here

File: src/plugins/data_table/data_table.js [hunk 16]
```diff
@@ -433,6 +433,16 @@ var LocalDataTable = (function() {
...
+        if (matches && matches[1]) {
+          cg.columnsConfig.forEach(function(currentObject){
+            if (currentObject.attr && Backdraft.Utils.toCSSClass(currentObject.attr).replace("column-","") === matches[1]) {
+              col = currentObject;
+            }
+          })
+        }
```
Comment (nburwell) [inline src/plugins/data_table/data_table.js]: wouldn't it be easier to follow the logic if the below line was the `else` block here? (and not always executed on L434 since we may just overwrite it..)

```
col = cg.columnConfigByTitle.attributes[title];
```

File: src/plugins/data_table/column_types/attr.js (hunk 13 above)
Comment (albertovillalobos) [inline src/plugins/data_table/column_types/attr.js]: I see, will do.

File: src/plugins/data_table/column_types/base.js (hunk 14 above)
Comment (albertovillalobos) [inline src/plugins/data_table/column_types/base.js]: Got it

File: src/plugins/data_table/data_table.js (hunk 16 above)
Comment (albertovillalobos) [inline src/plugins/data_table/data_table.js]: On it

**Comment by Vicentecarrillo** [comment]
closing this and opening PR to backdraft integration branch

File: src/plugins/data_table/data_table.js (hunk 16 above)
Comment (albertovillalobos) [inline src/plugins/data_table/data_table.js]: Addressed

File: src/plugins/data_table/column_types/base.js (hunk 15 above)
Comment (albertovillalobos) [inline src/plugins/data_table/column_types/base.js]: Addressed

File: src/plugins/data_table/column_types/base.js (hunk 14 above)
Comment (albertovillalobos) [inline src/plugins/data_table/column_types/base.js]: Addressed

File: src/plugins/data_table/column_types/attr.js (hunk 13 above)
Comment (albertovillalobos) [inline src/plugins/data_table/column_types/attr.js]: Addressed
